internal/consumer/dlq.go

```diff
@@ -36,97 +38,148 @@ type (
 	// can take a message and put them into some sort
 	// of error queue for later processing
 	DLQ interface {
-		Close()
-		// Add adds the given message to DLQ
+		// Start the DLQ producer
+		Start() error
+		// Stop the DLQ producer and close resources it holds.
+		Stop()
+		// Add adds the given message to DLQ.
+		// This is a synchronous call and will block until sending is successful.
 		Add(m kafka.Message) error
 	}
-	dlqImpl struct {
-		topic       string
-		producer    sarama.SyncProducer
-		retryPolicy backoff.RetryPolicy
-		tally       tally.Scope
-		logger      *zap.Logger
-	}
+
 	dlqNoop struct{}
-)
 
-const (
-	dlqRetryInitialInterval = 200 * time.Millisecond
-	dlqRetryMaxInterval     = 10 * time.Second
-)
+	// bufferedErrorTopic is a client side abstraction for an error topic on the Kafka cluster.
+	// This library uses the error topic as a base abstraction for retry and dlq topics.
+	//
+	// bufferedErrorTopic internally batches messages, so calls to Add may block until the internal
+	// buffer is flushed and a batch of messages is send to the cluster.
+	bufferedErrorTopic struct {
+		topic    string
+		cluster string
+		producer sarama.AsyncProducer
 
-// NewDLQ returns a DLQ writer for writing to a specific DLQ topic
-func NewDLQ(topic, cluster string, producer sarama.SyncProducer, scope tally.Scope, logger *zap.Logger) DLQ {
-	return &dlqImpl{
-		topic:       topic,
-		producer:    producer,
-		retryPolicy: newDLQRetryPolicy(),
-		tally:       scope,
-		logger:      logger.With(zap.String("topic", topic), zap.String("cluster", cluster)),
+		stopC     chan struct{}
+		doneC   chan struct{}
+
+		scope     tally.Scope
+		logger    *zap.Logger
+		lifecycle *util.RunLifecycle
 	}
-}
+)
 
-// Add blocks until successfully enqueuing the given
-// message into the error queue
-func (d *dlqImpl) Add(m kafka.Message) error {
-	sm, err := d.newSaramaMessage(m)
+// NewDLQ returns a DLQ writer for writing to a specific DLQ topic
+func NewDLQ(topic, cluster string, client sarama.Client, scope tally.Scope, logger *zap.Logger) (DLQ, error) {
+	asyncProducer, err := sarama.NewAsyncProducerFromClient(client)
```

Comment:
you might also want to configure things like batch_size_threshold / flush_frequency etc here rather than relying on defaults from sarama async producer

Reply:
Set it in buildSaramaConfig in the builder.